extern/storage-sealing/input.go

```diff
@@ -0,0 +1,405 @@
+			// todo check deal start deadline (configurable)
+
+			sid := m.minerSectorID(sector.SectorNumber)
+			m.assignedPieces[sid] = append(m.assignedPieces[sid], cid)
+
+			return ctx.Send(SectorAddPiece{})
+		},
+	}
+
+	go func() {
+		defer m.inputLk.Unlock()
+		if err := m.updateInput(ctx.Context(), sector.SectorType); err != nil {
+			log.Errorf("%+v", err)
+		}
+	}()
+
+	return nil
+}
+
+func (m *Sealing) handleAddPiece(ctx statemachine.Context, sector SectorInfo) error {
+	ssize, err := sector.SectorType.SectorSize()
+	if err != nil {
+		return err
+	}
+
+	res := SectorPieceAdded{}
+
+	m.inputLk.Lock()
+
+	pending, ok := m.assignedPieces[m.minerSectorID(sector.SectorNumber)]
+	if ok {
+		delete(m.assignedPieces, m.minerSectorID(sector.SectorNumber))
+	}
+	m.inputLk.Unlock()
+	if !ok {
+		// nothing to do here (might happen after a restart in AddPiece)
+		return ctx.Send(res)
+	}
+
+	var offset abi.UnpaddedPieceSize
+	pieceSizes := make([]abi.UnpaddedPieceSize, len(sector.Pieces))
+	for i, p := range sector.Pieces {
+		pieceSizes[i] = p.Piece.Size.Unpadded()
+		offset += p.Piece.Size.Unpadded()
+	}
+
+	maxDeals, err := getDealPerSectorLimit(ssize)
+	if err != nil {
+		return xerrors.Errorf("getting per-sector deal limit: %w", err)
+	}
+
+	for i, piece := range pending {
+		m.inputLk.Lock()
+		deal, ok := m.pendingPieces[piece]
+		m.inputLk.Unlock()
+		if !ok {
+			return xerrors.Errorf("piece %s assigned to sector %d not found", piece, sector.SectorNumber)
+		}
+
+		if len(sector.dealIDs())+(i+1) > maxDeals {
+			// todo: this is rather unlikely to happen, but in case it does, return the deal to waiting queue instead of failing it
+			deal.accepted(sector.SectorNumber, offset, xerrors.Errorf("too many deals assigned to sector %d, dropping deal", sector.SectorNumber))
+			continue
+		}
+
+		pads, padLength := ffiwrapper.GetRequiredPadding(offset.Padded(), deal.size.Padded())
+
+		if offset.Padded()+padLength+deal.size.Padded() > abi.PaddedPieceSize(ssize) {
+			// todo: this is rather unlikely to happen, but in case it does, return the deal to waiting queue instead of failing it
+			deal.accepted(sector.SectorNumber, offset, xerrors.Errorf("piece %s assigned to sector %d with not enough space", piece, sector.SectorNumber))
+			continue
+		}
+
+		offset += padLength.Unpadded()
+
+		for _, p := range pads {
+			ppi, err := m.sealer.AddPiece(sectorstorage.WithPriority(ctx.Context(), DealSectorPriority),
+				m.minerSector(sector.SectorType, sector.SectorNumber),
+				pieceSizes,
+				p.Unpadded(),
+				NewNullReader(p.Unpadded()))
+			if err != nil {
+				err = xerrors.Errorf("writing padding piece: %w", err)
+				deal.accepted(sector.SectorNumber, offset, err)
+				return ctx.Send(SectorAddPieceFailed{err})
+			}
+
+			pieceSizes = append(pieceSizes, p.Unpadded())
+			res.NewPieces = append(res.NewPieces, Piece{
+				Piece: ppi,
+			})
+		}
+
+		time.Sleep(1 * time.Second) // TODO: deal tests are unhappy without this
```

Comment:
It looks like markets are unhappy when we go from handing the piece to storage to the sector being [pre]committed - TestAPIDealFlow will reliably hang without this sleep

Example logs:
* Hangs (without this sleep) -> https://gist.github.com/magik6k/40e4c922435aaf78788537892fb8b193
* Ok (without this sleep) -> https://gist.github.com/magik6k/94fc89b602c57b20fad11700bffeada8

Looking at those it looks like some events arrive/are processed in wrong order:

```
2021-01-20T22:42:29.875+0100	INFO	markets	loggers/loggers.go:20	storage event	{"name": "ProviderEventDealPublished", "proposal CID": "bafyreiacufjlcrs4qetkcgz35septcseaqaleh3u353ggpdq4otkbmx3em", "state": "StorageDealStaged", "message": ""}
2021-01-20T22:42:29.875+0100	INFO	sectors	storage-sealing/input.go:220	Adding piece for deal 2 (publish msg: bafy2bzacealjen3fcckueo74mvqklm3oi4dehkipjpwwy4kdkucaqyq63p34w)
2021-01-20T22:42:29.875+0100	INFO	sectors	storage-sealing/input.go:396	Creating sector	{"number": "3", "type": "deal", "proofType": 0}
2021-01-20T22:42:29.875+0100	WARN	sbmock	mock/mock.go:75	Add piece: {{1000 3} 0} 2032 0
2021-01-20T22:42:29.875+0100	WARN	sbmock	mock/mock.go:85	Generated Piece CID: baga6ea4seaqkblkqstq56cdkxshgrkw4jsagbidi5gfecs2x2xxiozf6rvcvqli
2021-01-20T22:42:29.875+0100	INFO	sectors	storage-sealing/input.go:197	deal added to a sector	{"deal": 2, "sector": "3", "piece": "baga6ea4seaqkblkqstq56cdkxshgrkw4jsagbidi5gfecs2x2xxiozf6rvcvqli"}
2021-01-20T22:42:29.876+0100	WARN	storageadapter	storageadapter/provider.go:138	New Deal: deal 2
2021-01-20T22:42:29.876+0100	INFO	sectors	storage-sealing/input.go:59	starting to seal deal sector	{"sector": "3", "trigger": "filled"}
2021-01-20T22:42:29.876+0100	INFO	markets	loggers/loggers.go:20	storage event	{"name": "ProviderEventDealHandedOff", "proposal CID": "bafyreiacufjlcrs4qetkcgz35septcseaqaleh3u353ggpdq4otkbmx3em", "state": "StorageDealAwaitingPreCommit", "message": ""}
2021-01-20T22:42:29.876+0100	INFO	sectors	storage-sealing/states_sealing.go:28	performing filling up rest of the sector...	{"sector": "3"}
2021-01-20T22:42:29.879+0100	INFO	sectors	storage-sealing/states_sealing.go:277	submitting precommit for sector 3 (deposit: 110909785477): 
2021-01-20T22:42:29.881+0100	WARN	fullnode	full/mpool.go:168	Push from ID address (f0100), adjusting to f3ssgc2tijccwa7lv6m4aaaiaqkd5ehjbetkeoewc5ia67h2braxspad34b4c4445atgnepqjiw7nso7brixiq
2021-01-20T22:42:29.884+0100	INFO	sectors	storage-sealing/states_sealing.go:295	Sector precommitted: 3
2021-01-20T22:42:29.892+0100	INFO	messagepool	messagepool/selection.go:89	message selection done	{"took": 0.000189423}
2021-01-20T22:42:29.958+0100	INFO	sectors	storage-sealing/states_sealing.go:315	precommit message landed on chain: 3
2021-01-20T22:42:29.967+0100	INFO	markets	loggers/loggers.go:20	storage event	{"name": "", "proposal CID": "bafyreiacufjlcrs4qetkcgz35septcseaqaleh3u353ggpdq4otkbmx3em", "state": "StorageDealSealing", "message": ""}
Deal state:  StorageDealCheckForAcceptance
2021-01-20T22:42:30.744+0100	INFO	markets	loggers/loggers.go:15	storage event	{"name": "ClientEventWaitForDealState", "proposal CID": "bafyreiacufjlcrs4qetkcgz35septcseaqaleh3u353ggpdq4otkbmx3em", "state": "StorageDealCheckForAcceptance", "message": "Provider state: StorageDealVerifyData"}
2021-01-20T22:42:30.749+0100	INFO	markets	loggers/loggers.go:15	storage event	{"name": "ClientEventDealAccepted", "proposal CID": "bafyreiacufjlcrs4qetkcgz35septcseaqaleh3u353ggpdq4otkbmx3em", "state": "StorageDealProposalAccepted", "message": ""}
2021-01-20T22:42:30.749+0100	INFO	storageadapter	storageadapter/client.go:140	DEAL ACCEPTED!
2021-01-20T22:42:30.754+0100	INFO	markets	loggers/loggers.go:15	storage event	{"name": "ClientEventFundsReleased", "proposal CID": "bafyreiacufjlcrs4qetkcgz35septcseaqaleh3u353ggpdq4otkbmx3em", "state": "StorageDealProposalAccepted", "message": ""}
2021-01-20T22:42:30.754+0100	INFO	markets	loggers/loggers.go:15	storage event	{"name": "ClientEventDealPublished", "proposal CID": "bafyreiacufjlcrs4qetkcgz35septcseaqaleh3u353ggpdq4otkbmx3em", "state": "StorageDealAwaitingPreCommit", "message": ""}
```

cc @dirkmc

Reply:
#5398 fixes this